### PR TITLE
Improved logging / timout fixes and code cleanup

### PR DESCRIPTION
**Pull request by marcschier**
Arg ptr checks and logs
    fix unit tests
    logging and timeouts
    Remove uamqp
    Sample improvements
    Add proxy port configuration to allow use of static proxy name
    Fix file permissions
    Add log upload to event hub

### CODE REVIEW

File: bld/src/proxyd/log.config [hunk 1]
```diff
@@ -5,6 +5,7 @@
 strict init=true
 buffer min=1024
 buffer max=0
+file perms=644
```
Comment (JuergenKosel) [inline bld/src/proxyd/log.config]: Is it okay, that every body could read the log?
Otherwise use 600.

Comment (marcschier) [inline bld/src/proxyd/log.config]: We had an issue that we could not scp the log files from the shared volume used by the proxy docker container.  While this is a sample config, when using command line to set the log file, 604 is used.  Maybe it makes sense to to allow the default 600 to be overridden via command line.

Comment (JuergenKosel) [inline bld/src/proxyd/log.config]: Then 644 makes more sense, then giving read permissions to others but not to the own group.

Comment (marcschier) [inline bld/src/proxyd/log.config]: Agreed.

**Comment by JuergenKosel** [comment]
I have just checked: This pull request fixes #14 (checked only for Linux) and #27.

**Comment by JuergenKosel** [comment]
The registration over http proxy is broken again:

```
=== Azure IoT Proxy Gateway === 
Fri Mar 31 16:00:56 2017 (pal_cred_init:21) [ERROR] Pal does not provide credential storage support. Ensure you properly safeguard all secrets on this device!
Version: 0.1.2.0.1.1-6-g032b0a7
Fri Mar 31 16:00:56 2017 (HTTPAPI_ExecuteRequest:552) [ERROR] curl_easy_perform() failed: Couldn't resolve proxy name

Fri Mar 31 16:00:56 2017 (HTTPAPI_ExecuteRequest:554) [ERROR] (result = HTTPAPI_OPEN_REQUEST_FAILED)
Fri Mar 31 16:00:56 2017 (HTTPAPIEX_ExecuteRequest:475) [ERROR] unable to recover sending to a working state
ERROR: er_connecting Install failed! Check parameters...
Operation failed.

```

**Comment by JuergenKosel** [comment]
I believe, the problem regarding the proxy is in the prx_host_init_from_command_line() function:

```
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".
[New Thread 0x7ffff3171700 (LWP 20266)]
Breakpoint 2 at 0x7ffff7b4f07a: file /home/klj/iot-gateway-proxy/src/prx_host.c, line 279.
(gdb) print val
$1 = 0x7fffffffdde0 "roxy.example.com:"
#1  0x00007ffff7b4f0a6 in prx_host_init_from_command_line (host=0x61f4a0, argc=15, argv=0x7fffffffe058) at /home/klj/iot-gateway-proxy/src/prx_host.c:280
280	                __prx_config_set(prx_config_key_proxy_host, buffer);
(gdb) print optarg
$2 = 0x7fffffffe45e "proxy.example.com:3128"

```

The first character of the proxy has been cut away. And the remaining string still contains the trailing ':'.

Maybe you could install a proxy, like squid ( https://packages.debian.org/jessie/squid3 ) on one PC in your network for testing.

**Comment by marcschier** [comment]
Closing, will test proxy first.

**Comment by JuergenKosel** [comment]
With the latest commits the registration over http-proxy works again.
But for the normal operation I have a segmentation fault:

```
Starting program: /tmp/azure/usr/local/bin/proxyd -C ~/oriole_build/connectstring -D /home/klj/iot-gateway-proxy/azure --version --proxy proxy.example.com:3128 --proxy-user me --proxy-pwd 'secret' --only-websocket
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".
[New Thread 0x7ffff3171700 (LWP 21689)]
[New Thread 0x7ffff2920700 (LWP 21690)]
(gdb) bt
#0  __strcat_sse2_unaligned () at ../sysdeps/x86_64/multiarch/strcat-sse2-unaligned.S:46
#1  0x00007ffff7b90686 in pal_wsworker_get_proxy_info (proxy_address=0x7ffff291fb60, port=0x7ffff291fb68) at /home/klj/iot-gateway-proxy/src/pal/pal_ws_lws.c:1072
#2  0x00007ffff7b90b45 in pal_wsworker_pool_attach (pool=0x649450, wsclient=0x7fffec001100) at /home/klj/iot-gateway-proxy/src/pal/pal_ws_lws.c:1179
#3  0x00007ffff7b91220 in pal_wsclient_connect (wsclient=0x7fffec001100) at /home/klj/iot-gateway-proxy/src/pal/pal_ws_lws.c:1388
#4  0x00007ffff7b8515f in xio_wsclient_open (handle=0x7fffec0009e0, on_io_open_complete=0x7ffff7badd69 <onOpenComplete>, on_io_open_complete_context=0x7fffec0008c0, on_bytes_received=0x7ffff7badf7f <onBytesReceived>, on_bytes_received_context=0x7fffec0008c0, on_io_error=0x7ffff7bae017 <onIoError>, on_io_error_context=0x7fffec0008c0) at /home/klj/iot-gateway-proxy/src/xio_ws.c:482
#5  0x00007ffff7ba5b2b in xio_open (xio=0x7fffec0009c0, on_io_open_complete=0x7ffff7badd69 <onOpenComplete>, on_io_open_complete_context=0x7fffec0008c0, on_bytes_received=0x7ffff7badf7f <onBytesReceived>, on_bytes_received_context=0x7fffec0008c0, on_io_error=0x7ffff7bae017 <onIoError>, on_io_error_context=0x7fffec0008c0) at /home/klj/iot-gateway-proxy/deps/shared/src/xio.c:88
#6  0x00007ffff7baf6b7 in mqtt_client_connect (handle=0x7fffec0008c0, xioHandle=0x7fffec0009c0, mqttOptions=0x7ffff291fe00) at /home/klj/iot-gateway-proxy/deps/umqtt/src/mqtt_client.c:896
#7  0x00007ffff7b6ac92 in io_mqtt_connection_reconnect (connection=0x663500) at /home/klj/iot-gateway-proxy/src/io_mqtt.c:1044
#8  0x00007ffff7b58069 in prx_scheduler_work (context=0x64a940) at /home/klj/iot-gateway-proxy/src/prx_sched.c:229
#9  0x00007ffff7ba983c in ThreadWrapper (threadInstanceArg=0x64ac70) at /home/klj/iot-gateway-proxy/deps/shared/adapters/threadapi_pthreads.c:34
#10 0x00007ffff72f6064 in start_thread (arg=0x7ffff2920700) at pthread_create.c:309
#11 0x00007ffff786762d in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:111
(gdb) up
#1  0x00007ffff7b90686 in pal_wsworker_get_proxy_info (proxy_address=0x7ffff291fb60, port=0x7ffff291fb68) at /home/klj/iot-gateway-proxy/src/pal/pal_ws_lws.c:1072
1072	        strcat(*proxy_address, user);
```

File: src/pal/pal_ws_lws.c [hunk 2]
```diff
@@ -1045,16 +1049,23 @@ static int32_t pal_wsworker_get_proxy_info(
...
+        + 1;  // plus null term
     
-    *proxy_address = (char*)crt_alloc(buf_len); // Use malloc for lws to free
+    // Use malloc for lws to free
+    *proxy_address = (char*)crt_alloc(buf_len); 
     if (!*proxy_address)
         return er_out_of_memory;
-    memset(*proxy_address, 0, buf_len); // Avoid use of uninitialized memory, issue #27
+
+    *proxy_address = 0;
```
Comment (JuergenKosel) [inline src/pal/pal_ws_lws.c]: Undoing this change fixes the segmentation fault, which I reported a few moments ago.
So the memset for the whole *proxy_address buffer is needed, not just the first character.

Comment (marcschier) [inline src/pal/pal_ws_lws.c]: Now it is getting embarrassing.  Apologies...

**Pull request merged**